### DISCUSSION TITLE
LittleFS mounted on RAM - STM32F429zi

### DISCUSSION THREAD

**Discussion by GeorgePriestner-Eaton** [zephyrproject-rtos/zephyr#54166] [Q&A] [answered]
Hi all,

Does anyone know if it is possible to mount LittleFs file system on RAM?

I have successfully mounted FatFs instance on RAM but it requires too much RAM space for our needs.

I'm using STM32F429zi board with 192KB SRAM.

Thank you.

**Comment by GeorgePriestner-Eaton** [comment]
@erwango Do you know if this is possible please? Thank you in advance!

**Reply by erwango** [reply]
@GeorgePriestner-Eaton Please provide details: Did you have a try ? If yes did you face issues ?

**Reply by GeorgePriestner-Eaton** [reply]
Yep I tried to mount littlefs on RAM with following configuration:

```
static int xdl_fat_mount_app_fs(struct fs_mount_t *mnt)
{
	int rc = 0U;

#if CONFIG_FAT_FILESYSTEM_ELM
	static FATFS fat_fs;

	mnt->type = FS_FATFS;
	mnt->fs_data = &fat_fs;
	if (IS_ENABLED(CONFIG_DISK_DRIVER_RAM)) {
		mnt->mnt_point = "/RAM:";
	}
	else
	{
		return -1;
	}
#elif CONFIG_FILE_SYSTEM_LITTLEFS
	static fs_littlefs lfs;

	mnt->type = FS_LITTLEFS;
	mnt->fs_data = &lfs;
	if (IS_ENABLED(CONFIG_DISK_DRIVER_RAM)) {
		mnt->mnt_point = "/RAM:";
	}
#endif

	rc = fs_mount(mnt);

	if( rc < 0 )
	{
		printk("File system - Mount error\n");
	}

	return rc;
}
```

I also attached my prj.conf file.


It runs to fs_mount() but then hangs as it attempts to access flash api. So it seems it's always attempting to configure the file system with flash. I was wondering if there is any CONFIG settings I am missing, to force file system usage on RAM.

[prj.zip](https://github.com/zephyrproject-rtos/zephyr/files/10535164/prj.zip)

**Comment by Laczen** [comment]
@GeorgePriestner-Eaton, there are several ways to achieve this:
a. Setup littlefs to directly use RAM, this is done in one of the tests of littlefs.
b. Use the flash simulator to simulate flash on ram and use the simulated flash interface,
c. Use the ram disk interface, then you need to enable `CONFIG_FS_LITTLEFS_BLK_DEV`

**Reply by GeorgePriestner-Eaton** [reply]
Hey @Laczen, thanks for your reply!

a. Could you please point me to the test which uses RAM for littlefs mount? I checked the samples/subsys/fs directory but I only found littlefs mounted on non-volatile memory. I also checked samples\subsys\usb\mass example but this again uses SPI flash.

b. & c. I'll look into these, thanks. Do you have any examples at all please?

I'm still relatively new to using Zephyr, so still working things out. Thank you!

**Reply by Laczen** [reply]
Regarding a. I was really convinced that I saw this in the past, but I can't find it anymore.

Regarding b. You would need to add an extra flash controller in the dts file (an example for the simulator can be found in native_posix.dts), as you already have a flash controller do not use the name flashcontroller0. On this flash controller you should define partitions (at least 1) and you should use that partition for the littlefs,

Regarding c. You can enable the ram disk backend by selecting `CONFIG_DISK_DRIVERS = y`, `CONFIG_DISK_DRIVER_RAM = y`, this will create a 96kB disk in ram (you can also change the size) named `RAM`. Then look at examples that use littlefs on a block device (don't forget to set `CONFIG_FS_LITTLEFS_BLK_DEV = y`.

**Reply by Laczen** [reply]
@GeorgePriestner-Eaton, have you been able to get littlefs on RAM working?

**Reply by GeorgePriestner-Eaton** [reply]
I have not had chance to look into the flash simulator yet but when I do I will let you know! Thank you

**Reply by GeorgePriestner-Eaton** [reply]
@Laczen There doesn't seem to be any documentation on the flash simulator, other than the CONFIG page... 👎 

https://docs.zephyrproject.org/2.7.4/reference/kconfig/CONFIG_FLASH_SIMULATOR.html

Regarding c.

**Reply by Laczen** [reply]
@GeorgePriestner-Eaton, that is why I added the comment to have a look at native_posix.dts. In general you will find little documentation on a driver implementation (e.g. spi_nor.c, flash_xxx.c, soc_flash_xxx.c), and this is no different for the flash_simulator.c. You should just add a flash_controller similar to how it is done in native_posix.dts (and enable it using `CONFIG_FLASH_SIMULATOR=y`). All the driver does is reserve some space in ram and provide routines to use them as flash. The configuration of the flash_simulator would probably be done differently nowadays (using memory area) as well as more documented in the driver code, but this driver was created long before the present possibilities were available.

**Reply by vachagan-petrosyan** [reply]
Hello everyone, I have a same problem and I'm looking for a solution. I have software that needs to be ported to Zephyr and it contains a lot of code that reads/writes to files. I'm interested in creating a littlefs file system on RAM or simulating flash in RAM. Have any of you found a solution to this issue?

**Reply by Laczen** [reply]
> Hello everyone, I have a same problem and I'm looking for a solution. I have software that needs to be ported to Zephyr and it contains a lot of code that reads/writes to files. I'm interested in creating a littlefs file system on RAM or simulating flash in RAM. Have any of you found a solution to this issue?

Hi @vachagan-petrosyan, it should be possible to use the flash simulator for this. What device are you using ?

**Reply by vachagan-petrosyan** [reply]
Thank you @Laczen for your reply,

I am using Zephyr and STM32L4S9, without any additional memory (QSPI, SPI, FMC etc).

**Reply by Laczen** [reply]
> Thank you @Laczen for your reply,
> 
> I am using Zephyr and STM32L4S9, without any additional memory (QSPI, SPI, FMC etc).

How much memory would you reserve for the ram filesystem ?

**Reply by Laczen** [reply, marked as answer]
Using the following (added to the dts) or as a overlay file (for a 32kB filesystem), and enable the flash simulator (`CONFIG_FLASH_SIMULATOR=y`):
```
/ {
	soc {
		flash_controller2: flash-controller@2 {
			compatible = "zephyr,sim-flash";
			reg = <0x00000000 DT_SIZE_K(32)>;

			#address-cells = <1>;
			#size-cells = <1>;
			erase-value = <0xff>;

			flash_sim0: flash_sim@0 {
				status = "okay";
				compatible = "soc-nv-flash";
				erase-block-size = <512>;
				write-block-size = <1>;
				reg = <0x00000000 DT_SIZE_K(32)>;

				partitions {
					compatible = "fixed-partitions";
					#address-cells = <1>;
					#size-cells = <1>;

					lfsram_partition: partition@0 {
						label = "lfsram";
						reg = <0x00000000 0x000008000>;
					};
				};
			};
		};
	};
};

/ {
	fstab {
		compatible = "zephyr,fstab";
		lfsram: lfsram {
			compatible = "zephyr,fstab,littlefs";
			mount-point = "/lfsram";
			partition = <&lfsram_partition>;
			automount;
			read-size = <16>;
			prog-size = <16>;
			cache-size = <64>;
			lookahead-size = <32>;
			block-cycles = <512>;
		};
	};
};
```

**Reply by vachagan-petrosyan** [reply]
@Laczen Thank you very much it solve my problem at once. 

Once again, thank you for your prompt and efficient assistance.

**Reply by Laczen** [reply]
> @Laczen Thank you very much it solve my problem at once.
> 
> Once again, thank you for your prompt and efficient assistance.

Could you mark the Q&A as answered, the original author seems no longer involved.

**Reply by GeorgePriestner-Eaton** [reply]
I will try your above suggestion too. Thanks @Laczen.

**Reply by Laczen** [reply]
@GeorgePriestner-Eaton, @vachagan-petrosyan, FYI I submitted #57380,  this should make it easier for other users and has a bonus that the fs will survive a reboot.

**Reply by vachagan-petrosyan** [reply]
Thank you! : )

**Comment by Laczen** [comment]
@GeorgePriestner-Eaton, is this still unanswered?

**Reply by dman82499** [reply]
For Method C, I believe there are still issues. When setting up disk access to use FATFS, everything appears to work, but when I try to use littlefs in the block device mode (I verified it was using the disk access api by tracing through with my debugger), it does not seem to work. Could anybody verify this. I also could not find the tests that were used. 

Edit: I see you have a repository called NV that contains a block based solution on zephyr, I will try that out

**Reply by Laczen** [reply]
@dman82499, method c should also work. Could you give some more detail one what doesn't work and how you have set it up.

**Reply by dman82499** [reply]
My apologies, it actually does work, but, the sample projects for both littlefs and mass are not configured correctly. For the current samples, the buffer for disk access calls with littlefs was too small for the default sector size (512), which is the usually the minimum in all disk driver implementations. This caused those a buffers to overflow when they would be inserted in a disk access call. Of course it's an easy fix, to just set the littlefs read, prog and cache sizes to numbers >= 512, but for someone new it took some debugging and awful uart crashes to figure it out. Perhaps some of the KConfigs in the zephyr repository could be edited such that they use the correct buffer sizes, when it is configured as a block system instead of flash area? Should be a pretty easy PR to make.
